File: internal/tui/knowledgegit.go
package tui

import (
	"context"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
	"unicode"

	"loom/internal/config"
)

// gitTimeout bounds every git invocation. Promote and reject run inside the
// fullscreen TUI, where a child that blocks — on an index lock, a credential or
// a signing prompt — is unrecoverable for the user.
const gitTimeout = 10 * time.Second

// gitReasonMax bounds the git failure text that reaches the status line, so a
// multi-line failure contributes only its head to a bar that already carries the
// destination path. It is not what keeps the line inside the terminal — the view
// clamps the composed status to the window width — and it bounds nothing that is
// written down: knowledge-git.log keeps git's whole output.
const gitReasonMax = 60

// gestureMessageMax bounds the commit subject, which is also the body of the
// knowledge-git log record.
const gestureMessageMax = 200

// errNoGitRepo marks a knowledge root that is not under version control, so the
// gesture can say the store isn't a repo rather than relay a git failure.
var errNoGitRepo = errors.New("knowledge root is not a git repo")

// errEnclosingRepo marks a knowledge root that is itself untracked but sits
// inside a repo — a git-managed home directory, a dotfiles checkout. It degrades
// exactly like errNoGitRepo but needs its own reason, because the user can
// falsify "not a git repo" by running git status in the store.
var errEnclosingRepo = errors.New("knowledge root is inside another git repo")

// knowledgeGitLogPath returns the canonical log file for knowledge-store
// commits that could not be recorded.
func knowledgeGitLogPath() string {
	return filepath.Join(config.Home(), "knowledge-git.log")
}

// runGit runs one git command in the knowledge store, returning its combined
// output. commit.gpgsign=false because a signing passphrase prompt would hang
// the TUI with no way to answer it.
func runGit(root string, args ...string) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), gitTimeout)
	defer cancel()
	full := append([]string{"-C", root, "-c", "commit.gpgsign=false"}, args...)
	out, err := exec.CommandContext(ctx, "git", full...).CombinedOutput()
	// CommandContext reports a killed child as a generic signal, so the
	// deadline that fired is visible only on the context — and a child killed
	// mid-prompt has printed nothing to attribute the failure to.
	if err != nil && ctx.Err() != nil {
		err = ctx.Err()
	}
	return strings.TrimSpace(string(out)), err
}

// gitCause pairs git's output with the error that ended the child, so the two
// failures that print nothing — a git that could not be run, a child killed by
// gitTimeout — still record a cause.
func gitCause(out string, err error) string {
	if out == "" {
		return err.Error()
	}
	return err.Error() + ": " + out
}

// commitKnowledge records the paths a gesture touched as one commit in the
// knowledge store. The commit is path-scoped — `git add` over those paths, then
// `commit --only` with the same pathspec — because the store's working tree is
// routinely dirty with untracked candidates and edits this gesture did not
// make, and a whole-tree commit would absorb them into the record. droppable
// names the paths whose record lives elsewhere, committed alongside the rest but
// abandoned rather than allowed to sink the commit; every other path is the
// record and fails loudly; a dropped path is named in knowledgeGitLogPath(),
// since the commit that lands is otherwise indistinguishable from any other.
// That record is written when the path leaves the pathspec, so it stands even
// when the commit it was headed for then fails. A root that is not itself a git
// repo yields errNoGitRepo. Failures carry git's whole output, because the useful part is
// rarely the first line — a rejected commit leads with "On branch main" and
// names the cause below it — and the caller, not this function, decides what a
// one-line status bar can show.
func commitKnowledge(paths, droppable []string, message string) error {
	root := KnowledgeRoot()
	top, err := runGit(root, "rev-parse", "--show-toplevel")
	if err != nil {
		// rev-parse fails both for a store that is not a repo and for a git we
		// could not run at all — missing from PATH, or killed by gitTimeout.
		// Only an exit status means git ran and answered; blaming the store's
		// layout for anything else would send the user to fix a healthy repo.
		var exitErr *exec.ExitError
		if !errors.As(err, &exitErr) {
			return fmt.Errorf("git rev-parse: %s", gitCause(top, err))
		}
		return fmt.Errorf("%w: %s", errNoGitRepo, gitCause(top, err))
	}
	// rev-parse walks up the tree, so a store that is merely *inside* a repo —
	// a git-managed home directory, a dotfiles checkout — resolves to that
	// ancestor and the commit would land in history the user never pointed us
	// at. The toplevel has to be the store itself.
	if !sameDir(top, root) {
		return fmt.Errorf("%w: enclosing repo at %s", errEnclosingRepo, top)
	}
	// A droppable path leaves the pathspec when git is ignoring it: its record
	// is elsewhere — the reject archive's decision is the log.md entry — so the
	// store's ignore rules are a storage policy there, while passing an ignored
	// path to `git add` is a fatal "paths are ignored" that would sink the whole
	// commit. This is not a general rule about ignored paths: a path the caller
	// did not declare droppable is the record, and an ignored one has to fail
	// loudly rather than yield a commit that omits it. check-ignore exits 0 only
	// when the path is ignored — 1 when git would track it, 128 on an error it
	// could not answer — so only a clean exit drops the path.
	kept := append([]string{}, paths...)
	for _, p := range droppable {
		if _, err := runGit(root, "check-ignore", "-q", "--", p); err == nil {
			appendKnowledgeGitLog(message + ": dropping ignored path " + shortenPath(p))
			continue
		}
		kept = append(kept, p)
	}
	// A path the gesture removed matters to git only if it was tracked; the
	// store carries uncommitted candidates, and passing one as a pathspec after
	// its removal is a fatal "did not match any files" that would sink the
	// whole commit.
	var pathspec []string
	for _, p := range kept {
		if _, err := os.Stat(p); err != nil {
			if _, err := runGit(root, "ls-files", "--error-unmatch", "--", p); err != nil {
				continue
			}
		}
		pathspec = append(pathspec, p)
	}
	if len(pathspec) == 0 {
		return errors.New("no tracked paths to commit")
	}
	if out, err := runGit(root, append([]string{"add", "--"}, pathspec...)...); err != nil {
		return fmt.Errorf("git add: %s", gitCause(out, err))
	}
	if out, err := runGit(root, append([]string{"commit", "--only", "-m", message, "--"}, pathspec...)...); err != nil {
		// A failed commit leaves the gesture staged, and the next commit a
		// human makes in the store would absorb it — the mirror of the
		// absorption the pathspec scoping exists to prevent. Unstaging also
		// drops any pre-gesture staging of these exact paths, which for a
		// just-written destination and a just-removed candidate is no real loss.
		runGit(root, append([]string{"reset", "-q", "--"}, pathspec...)...)
		return fmt.Errorf("git commit: %s", gitCause(out, err))
	}
	return nil
}

// recordKnowledgeCommit commits a gesture and returns "" on success or a short
// reason on failure. A failure is never silent: the failure is appended to
// knowledgeGitLogPath() in full, and its head handed back for the status line,
// which is gone by the next keystroke.
func recordKnowledgeCommit(paths, droppable []string, message string) string {
	message = sanitizeRecord(message)
	err := commitKnowledge(paths, droppable, message)
	if err == nil {
		return ""
	}
	reason := recordKnowledgeFailure(message, err)
	// Both no-repo shapes degrade the same way; the sentinel text is the whole
	// status-line reason, since the enclosing path detail belongs in the log.
	if errors.Is(err, errNoGitRepo) {
		return errNoGitRepo.Error()
	}
	if errors.Is(err, errEnclosingRepo) {
		return errEnclosingRepo.Error()
	}
	return reason
}

// recordKnowledgeFailure writes one gesture failure to knowledgeGitLogPath() in
// full and returns its head for the status line, which is gone by the next
// keystroke. Shared by the commit and by the log.md append reject records into,
// so both failures reach the same two places. The message is sanitized here —
// idempotent on text a caller already sanitized — so the bound a record shares
// with the commit subject is applied once, at the boundary both failures pass
// through, rather than in each caller; the flattening half of the invariant
// sits lower still, in appendKnowledgeGitLog.
func recordKnowledgeFailure(message string, err error) string {
	message = sanitizeRecord(message)
	// Flattened but not bounded: the log is the debugging mechanism for this
	// gesture, so it keeps every line the failure produced as one record.
	appendKnowledgeGitLog(message + ": " + flattenRecord(err.Error()))
	return shortGit(err.Error())
}

// sanitizeRecord flattens a gesture message into a single bounded record — the
// commit subject, which is also the body of the knowledge-git log line. The
// bound counts runes, not bytes: a non-ASCII id near the limit would otherwise
// be cut mid-rune, writing invalid UTF-8 into the very record flattenRecord
// exists to keep readable.
func sanitizeRecord(message string) string {
	flat := []rune(flattenRecord(message))
	if len(flat) <= gestureMessageMax {
		return string(flat)
	}
	return string(flat[:gestureMessageMax-1]) + "…"
}

// flattenRecord maps every rune that could break a one-line record onto a space:
// control characters, the Unicode format characters (bidi overrides and
// isolates, which reorder a record without appearing in it) and the line and
// paragraph separators. The text it guards names an artifact id and scope taken
// from candidate frontmatter the LLM extractor wrote out of a session
// transcript, so an unfiltered rune would forge a commit-message trailer, an
// extra knowledge-git.log line, or a misleading rendering of either —
// corrupting the audit trail these records exist to be.
func flattenRecord(s string) string {
	flat := strings.Map(func(r rune) rune {
		if unicode.IsControl(r) || unicode.Is(unicode.Cf, r) || r == '\u2028' || r == '\u2029' {
			return ' '
		}
		return r
	}, s)
	return strings.TrimSpace(flat)
}

// appendKnowledgeGitLog appends one timestamped line to the knowledge-git log.
// The line is flattened here rather than in each caller, so no caller can write
// a record that forges a second one; flattenRecord is idempotent, so the callers
// that flatten their own variable part first are unaffected. Defence in depth
// rather than a live hole: no extractor-written text reaches this unflattened
// today, since CANDIDATE_ID_RE gates the candidate filename.
func appendKnowledgeGitLog(line string) {
	p := knowledgeGitLogPath()
	if err := os.MkdirAll(filepath.Dir(p), 0o700); err != nil {
		return
	}
	f, err := os.OpenFile(p, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o600)
	if err != nil {
		return
	}
	defer f.Close()
	fmt.Fprintf(f, "%s %s\n", time.Now().Format(time.RFC3339), flattenRecord(line))
}

// sameDir reports whether two paths name the same directory. Both sides are
// resolved first: git reports an absolute physical path while KnowledgeRoot()
// hands back LOOM_KNOWLEDGE_ROOT verbatim, which may be relative, and on macOS
// a store under /var/folders comes back as /private/var/folders — the raw
// strings would never compare equal either way.
func sameDir(a, b string) bool {
	ra, err := resolveDir(a)
	if err != nil {
		return false
	}
	rb, err := resolveDir(b)
	if err != nil {
		return false
	}
	return ra == rb
}

// resolveDir makes a path absolute and then resolves its symlinks.
func resolveDir(p string) (string, error) {
	abs, err := filepath.Abs(p)
	if err != nil {
		return "", err
	}
	return filepath.EvalSymlinks(abs)
}

// shortGit collapses a failure to its first line, bounded, so it fits the
// one-line status bar. It is the display boundary only — callers log the whole
// text before shortening it.
func shortGit(out string) string {
	first, _, _ := strings.Cut(out, "\n")
	return truncate(first, gitReasonMax)
}

File: extractors/knowledge_git.py
#!/usr/bin/env python3
"""Commit the knowledge store's own git repo for what an extraction run wrote.

Python counterpart to internal/tui/knowledgegit.go, which does the same for the
promote and reject gestures. The extractor is the store's highest-volume writer:
without this every sweep leaves candidate files and a log.md entry as
working-tree state only, and anything reading the repo's history sees the corpus
stand still.

The unit is the run, not the file — one commit per sweep, matching the one entry
log.md already records — so a backlog of sweeps reads back as a list of
extraction events rather than a file-by-file diff.
"""
from __future__ import annotations

import os
import subprocess
import unicodedata
from datetime import datetime
from pathlib import Path

# GIT_TIMEOUT bounds every git invocation. This runs unattended on the
# LaunchAgent, so a child that blocks — on an index lock, a credential or a
# signing prompt — would wedge the sweep with nobody there to answer it; there
# is no interactive TUI to starve either, so the bound is looser than the TUI's.
GIT_TIMEOUT = 30

# REASON_MAX bounds the git failure text handed back to the caller, which prints
# it as one [extract] line. It bounds nothing that is written down:
# knowledge-git.log keeps git's whole output.
REASON_MAX = 60

# MESSAGE_MAX bounds the commit subject, which is also the body of the
# knowledge-git log record.
MESSAGE_MAX = 200


class NoGitRepoError(Exception):
    """A knowledge root that is not under version control, so the run can say
    the store isn't a repo rather than relay a git failure."""

    reason = "knowledge root is not a git repo"


class EnclosingRepoError(Exception):
    """A knowledge root that is itself untracked but sits inside a repo — a
    git-managed home directory, a dotfiles checkout. It degrades exactly like
    NoGitRepoError but needs its own reason, because the user can falsify "not a
    git repo" by running git status in the store."""

    reason = "knowledge root is inside another git repo"


def knowledge_git_log_path() -> Path:
    """Canonical log file for knowledge-store commits that could not be
    recorded. Mirrors internal/config.Home()."""
    home = os.environ.get("LOOM_HOME")
    return (Path(home) if home else Path.home() / ".loom") / "knowledge-git.log"


def _run_git(root: Path, *args: str) -> tuple[str, int | str]:
    """Run one git command in the knowledge store, returning its combined output
    and either git's exit status or, when git could not be run at all, a string
    describing why. Callers must keep the two apart: only an exit status means
    git ran and answered. Signing and hooks are both pinned off for the same
    reason: a sweep runs unattended, so a signing passphrase prompt or a hook
    that blocks would wedge it until GIT_TIMEOUT on every run. The store is a
    data store loom's bootstrap creates, not a project checkout whose hooks
    anyone meant to run here. core.hooksPath=/dev/null names a location git
    finds no hooks in — it is a file rather than a directory, and git treats
    that as no hooks rather than an error. The encoding is pinned rather than
    left to the locale: the LaunchAgent runs under C/POSIX, where a non-ASCII
    candidate name in git's output would raise UnicodeDecodeError — neither
    OSError nor TimeoutExpired, so it would escape this function's handlers and
    be recorded as a decode error instead of the git failure it hid."""
    full = ["git", "-C", str(root), "-c", "commit.gpgsign=false",
            "-c", "core.hooksPath=/dev/null", *args]
    try:
        proc = subprocess.run(full, stdout=subprocess.PIPE,
                              stderr=subprocess.STDOUT,
                              encoding="utf-8", errors="replace",
                              timeout=GIT_TIMEOUT)
    except OSError as exc:
        # A missing binary: nothing ran, so there is nothing it printed.
        return "", f"{type(exc).__name__}: {exc}"
    except subprocess.TimeoutExpired as exc:
        # A child killed at the deadline often named its own cause before it
        # blocked — a git waiting on index.lock says so — so keep what it wrote,
        # which arrives as bytes (or None) whatever the decoding args say.
        out = exc.output or b""
        if isinstance(out, bytes):
            out = out.decode(errors="replace")
        return out.strip(), f"{type(exc).__name__}: {exc}"
    return proc.stdout.strip(), proc.returncode


def _cause(out: str, err: int | str) -> str:
    """Pair git's output with what ended the child, so the two failures that
    print nothing — a git that could not be run, a child killed at GIT_TIMEOUT —
    still record a cause."""
    detail = f"exit status {err}" if isinstance(err, int) else err
    if not out:
        return detail
    return f"{detail}: {out}"


def commit_knowledge(root: Path, paths: list[Path], message: str) -> None:
    """Record the paths one extraction run wrote as a single commit in the
    knowledge store, raising on failure. The commit is path-scoped — `git add`
    over those paths, then `commit --only` with the same pathspec — because the
    store's working tree is routinely dirty with untracked candidates and edits
    this run did not make, and a whole-tree commit would absorb them into the
    record. Failures carry git's whole output, because the useful part is rarely
    the first line — a rejected commit leads with "On branch main" and names the
    cause below it — and the caller decides what one printed line can show."""
    top, err = _run_git(root, "rev-parse", "--show-toplevel")
    if err != 0:
        # rev-parse fails both for a store that is not a repo and for a git we
        # could not run at all. Blaming the store's layout for anything but an
        # exit status would send the user to fix a healthy repo.
        if not isinstance(err, int):
            raise RuntimeError(f"git rev-parse: {_cause(top, err)}")
        raise NoGitRepoError(f"{NoGitRepoError.reason}: {_cause(top, err)}")
    # rev-parse walks up the tree, so a store that is merely *inside* a repo
    # resolves to that ancestor and the commit would land in history nobody
    # pointed us at. The toplevel has to be the store itself. Both sides are
    # resolved first: git reports an absolute physical path, while the root
    # arrives as LOOM_KNOWLEDGE_ROOT verbatim, and on macOS a store under
    # /var/folders comes back as /private/var/folders — the raw strings would
    # never compare equal either way.
    if Path(top).resolve() != Path(root).resolve():
        raise EnclosingRepoError(f"{EnclosingRepoError.reason}: enclosing repo at {top}")
    # A path matters to git only if it exists or was tracked; the store carries
    # uncommitted candidates, and naming one that is neither is a fatal "did not
    # match any files" that would sink the whole commit.
    pathspec = []
    for p in paths:
        # Resolved for the same reason the toplevel is: git runs with -C root,
        # so a path the caller built against the process cwd — a relative
        # LOOM_KNOWLEDGE_ROOT gives relative paths — would be re-resolved
        # against the store, match nothing, and sink the commit.
        full = str(Path(p).resolve())
        if not Path(full).exists():
            _, err = _run_git(root, "ls-files", "--error-unmatch", "--", full)
            if err != 0:
                continue
        pathspec.append(full)
    if not pathspec:
        raise RuntimeError("no tracked paths to commit")
    out, err = _run_git(root, "add", "--", *pathspec)
    if err != 0:
        raise RuntimeError(f"git add: {_cause(out, err)}")
    out, err = _run_git(root, "commit", "--only", "-m", message, "--", *pathspec)
    if err != 0:
        # A failed commit leaves the run staged, and the next commit anyone
        # makes in the store would absorb it — the mirror of the absorption the
        # pathspec scoping exists to prevent.
        _run_git(root, "reset", "-q", "--", *pathspec)
        raise RuntimeError(f"git commit: {_cause(out, err)}")


def record_knowledge_commit(root: Path, paths: list[Path], message: str) -> str:
    """Commit one extraction run and return "" on success or a short reason on
    failure. Every exception is caught, anticipated or not: an unattended sweep
    must not fail the extraction because the commit failed. A failure is never
    silent — it is appended to knowledge_git_log_path() in full, and its head
    handed back for the caller to print."""
    message = sanitize_record(message)
    try:
        commit_knowledge(root, paths, message)
        return ""
    except Exception as exc:
        # Flattened but not bounded: the log is the debugging mechanism for
        # these commits, so it keeps every line the failure produced.
        _append_knowledge_git_log(f"{message}: {_flatten_record(str(exc))}")
        # Both no-repo shapes degrade the same way; the sentinel text is the
        # whole printed reason, since the enclosing path belongs in the log.
        if isinstance(exc, (NoGitRepoError, EnclosingRepoError)):
            return type(exc).reason
        return _short_git(str(exc))


def sanitize_record(message: str) -> str:
    """Flatten a run's message into a single bounded record — the commit
    subject, which is also the body of the knowledge-git log line. The subject
    names a scope resolved from --scope/resolve_project and a session id read
    out of the input file's frontmatter, so this is defence in depth at the
    boundary that writes the record rather than a check on any one caller."""
    flat = _flatten_record(message)
    # Python slices by code point, so a non-ASCII id near the limit is never cut
    # mid-character into the very record _flatten_record keeps readable.
    if len(flat) <= MESSAGE_MAX:
        return flat
    return flat[:MESSAGE_MAX - 1] + "…"


def _flatten_record(s: str) -> str:
    """Map every character that could break a one-line record onto a space:
    control characters (Cc), the Unicode format characters (Cf — bidi overrides
    and isolates, which reorder a record without appearing in it) and the line
    and paragraph separators (Zl, Zp). Unfiltered, one of these would forge a
    commit-message trailer, an extra knowledge-git.log line, or a misleading
    rendering of either — corrupting the audit trail these records exist to be."""
    return "".join(
        " " if unicodedata.category(c) in ("Cc", "Cf", "Zl", "Zp") else c
        for c in s
    ).strip()


def _append_knowledge_git_log(line: str) -> None:
    """Append one timestamped line to the knowledge-git log. A log that cannot
    be written is not worth failing a sweep over — and this runs inside
    record_knowledge_commit's handler, so anything escaping here would fail the
    extraction over a failed commit, which is the one outcome that path exists
    to prevent. The guard is Exception rather than OSError because resolving the
    default root raises RuntimeError when there is no home directory to find,
    where internal/config.Home() falls back instead of raising."""
    path = knowledge_git_log_path()
    try:
        # Modes mirror the Go writer's: the TUI gestures append to this same
        # file, and whichever side creates it sets what the other inherits.
        path.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
        fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_APPEND, 0o600)
        with os.fdopen(fd, "a") as f:
            f.write(f"{datetime.now().astimezone().replace(microsecond=0).isoformat()} {line}\n")
    except Exception:
        return


def _short_git(out: str) -> str:
    """Collapse a failure to its first line, bounded, so it fits one printed
    line. This is the display boundary only — the whole text is logged first."""
    first = out.split("\n", 1)[0]
    if len(first) <= REASON_MAX:
        return first
    return first[:REASON_MAX - 1] + "…"
